Allow deletion of teams.

peter-fogg opened the pull request:
## [TNL-1915](https://openedx.atlassian.net/browse/TNL-1915)
Allows deletion of teams by privileged users, using the standard Teams definition of privileged. Team memberships are permanently deleted along with teams.

### Sandbox
- [x] http://peter-fogg.sandbox.edx.org/courses/course-v1:TeamsX+T101+2015/teams/#my-teams

### Testing
- [x] i18n
- [x] RTL
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
Please check the appropriate box after all relevant reviewers have finished and given the :+1:.

- [x] Code review: @cahrens 
- [x] Code review: @andy-armstrong 
- [x] UI strings: @catong 
- [x] Product review: @explorerleslie 
- FYI @cptvitamin @andywaldrop @marcotuts 

### Post-review
- [x] Squash commits

andy-armstrong commented:
@peter-fogg @explorerleslie @catong I'm thinking thta the wording on the delete modal should be changed to sound more permanent. It currently says:

"Deleting a team removes it from the team listing view, and removes members from the team as well."

This was true when we were just marking it as inactive, but now we are completely purging it from the system. We probably should also mention that the team discussion is no longer accessible but is not actually removed.

andy-armstrong commented:
@peter-fogg @andywaldrop The "Close" button in the delete modal is very hard to read as yellow text on a gray background:

![image](https://cloud.githubusercontent.com/assets/5985072/9607521/a5534c1e-5097-11e5-8061-1fcf59686092.png)

Also, why is the "Close" rather than "Cancel"? I don't remember using such a label in other modals.

andy-armstrong commented on common/test/acceptance/pages/lms/teams.py:
```diff
@@ -230,9 +235,9 @@ def sort_teams_by(self, sort_order):
 
 
-class CreateOrEditTeamPage(CoursePage, FieldsMixin):
+class CreateEditDeleteTeamPage(CoursePage, FieldsMixin):
```
Nit: This is quite a mouthful. Could it be called the TeamEditorPage or TeamManagementPage?

peter-fogg replied:
Changed this to `TeamManagementPage`.

peter-fogg commented:
jenkins run all

@andy-armstrong @explorerleslie @catong I'm also not a fan of the wording; I just took what was in the wireframes and I decided to wait until review to figure out better copy.

The style issues are a temporary sandbox problem -- they should be fixed in about five minutes.

peter-fogg commented:
jenkins run all

andy-armstrong commented on common/test/acceptance/pages/lms/teams.py:
```diff
@@ -281,6 +286,20 @@ def cancel_team(self):
+        confirm_prompt(self, require_notification=False)
+        self.wait_for_ajax()
+
+    def cancel_delete_team_dialog(self):
```
Nit: I probably would call ``confirm_prompt`` directly from the test and not have this helper method. It seems heavyweight to have to add such a method for every modal that we might show on a page.

peter-fogg replied:
Done.

andy-armstrong commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -937,6 +948,76 @@ def test_user_can_cancel_the_team_creation(self):
+        self.team_page = TeamPage(self.browser, self.course_id, team=self.team)
+        self.team_page.visit()
+
+    def test_cancel_delete(self):
```
Maybe add a test that a non-staff user does not see the delete button? Or will that be covered by Eric's instructor toolbar PR?

peter-fogg replied:
I decided not to test privileges too aggressively, since non-privileged users shouldn't see the instructor toolbar. I think that's best tested in Eric's PR.

cahrens replied:
Currently @efischer19 doesn't have any tests in his PR.

efischer19 replied:
Yeah, Since they overlap so much I wasn't sure if we wanted to test each part of TNL-1913 and 1915 separately as they were made, or all together in one suite of tests.

I can add the types of tests being discussed here to https://github.com/edx/edx-platform/pull/9520, I agree that seems like a sensible place to test privileged access.

andy-armstrong commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -937,6 +948,76 @@ def test_user_can_cancel_the_team_creation(self):
+        self.create_edit_delete_team_page.wait_for_page()
+        self.create_edit_delete_team_page.delete_team_button.click()
+        self.create_edit_delete_team_page.cancel_delete_team_dialog()
+        self.assertTrue(self.create_edit_delete_team_page.is_browser_on_page())
```
Wouldn't this be true even if the cancel did nothing? Maybe you should verify that if you refresh the page that the team is still there. You could also verify that the modal has gone away, although maybe that should be a given.

peter-fogg replied:
I've changed the test to refresh the page and verify that the team is present.

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/views/instructor_tools.js:
```diff
@@ -25,14 +29,46 @@
-                    alert("You clicked the button!");
-                    //placeholder; will route to delete team page
+                    var self = this;
+                    var msg = new PromptView.Warning({
```
I recommend using ``ViewUtils.confirmThenRunOperation`` instead of constructing the modal yourself.

peter-fogg replied:
:trophy: I didn't know that existed.

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/views/instructor_tools.js:
```diff
@@ -25,14 +29,46 @@
+
+                handleDelete: function () {
+                    var self = this;
+                    this.team.destroy().then(function () {
```
I'm thinking that you should have error handling in case this team can't be deleted for some reason. We could rely on the general purposes AJAX error handling that we have to do in a subsequent story, but I'm thinking we could give more targeted feedback here.

peter-fogg replied:
I agree that this should have some sort of feedback if the deletion can't be completed. I'm not sure what the best way to display that feedback is, though -- we have the standard system notifications, but we've also got a teams-specific error pattern for things like invalid URL fragments. @frrrances, @marcotuts?

marcotuts replied:
I don't think we have a wireframe that conveys a deletion error. I'm fine with exploring the use of the teams error message on the topic page to convey deletion error (or maybe even deletion confirmation.) That being said, if possible this might be better handled as a follow-on story to make sure we're doing these messages as we'd like to.  I'll defer to @explorerleslie on what she'd like to do here.

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/views/instructor_tools.js:
```diff
@@ -25,14 +29,46 @@
+                handleDelete: function () {
+                    var self = this;
+                    this.team.destroy().then(function () {
+                        self.teamEvents.trigger('teams:update', {
```
I think you need to update ``collections/topic.js`` so that it marks itself stale on team deletion. It probably deserves a test too, that the team count is correctly updated on the topics page.

peter-fogg replied:
I'll add a Bok Choy test for the team count, and a Jasmine test for the topic collection.

andy-armstrong commented on lms/djangoapps/teams/views.py:
```diff
@@ -58,6 +60,9 @@
 MAXIMUM_SEARCH_SIZE = 100000
 
 
```
Nit: you don't need two lines here.

peter-fogg replied:
Fixed.

andy-armstrong commented on lms/djangoapps/teams/views.py:
```diff
@@ -58,6 +60,9 @@
+log = logging.getLogger(__name__)
+
+
 class TeamsDashboardView(View):
```
Nit: you didn't name this, but I'm not sure why this is called a dashboard view. We don't have the terminology of a dashboard anywhere...

andy-armstrong commented on lms/djangoapps/teams/views.py:
```diff
@@ -541,6 +562,15 @@ def get_queryset(self):
+        team = get_object_or_404(CourseTeam, team_id=team_id)
+        self.check_object_permissions(request, team)
+        # Note: also deletes all team memberships associated with this team
+        log.info('user %d deleted team %s', request.user.id, team_id)
```
Nit: put the log after the call to ``team.delete`` so that we don't log if the delete fails.

peter-fogg replied:
Done.

andy-armstrong commented on lms/djangoapps/teams/views.py:
```diff
@@ -541,6 +562,15 @@ def get_queryset(self):
+        self.check_object_permissions(request, team)
+        # Note: also deletes all team memberships associated with this team
+        log.info('user %d deleted team %s', request.user.id, team_id)
+        team.delete()
```
I wonder if there are any exceptions you should be catching here. The permission checks should ensure that the delete can go through, but there is a window where something might have changed (e.g. another user just deleted it before this code hits). Maybe it is okay if it returns a server error in that case, but it might be clearer to return a more accurate return code.

peter-fogg replied:
I don't think there are any exceptions that need to be covered, actually. Django will throw an error if an object has its `delete` method called multiple times, but a separate request will get a separate object, so no error will be thrown. MySQL doesn't care if you try to delete nonexistent data.

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/views/instructor_tools.js:
```diff
@@ -25,14 +29,46 @@
 
                 deleteTeam: function (event) {
```
Please add Jasmine tests for team deletion:
1. Can delete a team
2. Can cancel deletion of a team
3. Handles an error when deleting a team

There is some controversy about when to write Jasmine versus Bok Choy, and you already have Bok Choy tests here. My opinion is that we want good Jasmine code coverage, and it isn't possible to fake out server errors etc in Bok Choy. We also should follow the testing pyramid of preferring unit tests over integration tests:

https://github.com/edx/edx-platform/blob/master/docs/en_us/internal/testing.rst

andy-armstrong replied:
Also add a Jasmine test that the topics page refreshes when a team deletion event is fired.

peter-fogg replied:
Added these tests.

andy-armstrong replied:
Thanks, these Jasmine tests look great.

andy-armstrong commented on lms/djangoapps/teams/views.py:
```diff
@@ -530,6 +537,20 @@ class TeamsDetailView(ExpandableFieldViewMixin, RetrievePatchAPIView):
+            team memberships associated with that team are also
+            deleted.
+
+            If the user is anonymous of inactive, a 401 is returned.
```
Typo: 'of inactive'

peter-fogg replied:
Fixed.

andy-armstrong commented on lms/djangoapps/teams/views.py:
```diff
@@ -530,6 +537,20 @@ class TeamsDetailView(ExpandableFieldViewMixin, RetrievePatchAPIView):
+
+            Only staff can delete teams. When a team is deleted, all
+            team memberships associated with that team are also
+            deleted.
```
Mention the 204 return code.

peter-fogg replied:
Done.

andy-armstrong commented:
@explorerleslie I just wanted to confirm the behavior when a user is on a team discussion page and then a staff member deletes the team. Currently, any users in the discussion can continue posting to the thread and won't see any issues. It is only when they refresh that the team page disappears from beneath them. @peter-fogg and I discussed this and it seems difficult to do anything any differently, but I just wanted to confirm that you are okay with this.

The only option I could see is to also delete the team discussion at the same time, in which case the user would see errors. This might be better from a staff user's perspective if say they were deleting a team due to legal concerns about the discussion. If I remember correctly, we discussed this and decided that just making the discussion inaccessible was all we needed to do.

andy-armstrong commented:
@peter-fogg this is looking great. I'm done with my first pass.

peter-fogg commented:
jenkins run lettuce

explorerleslie commented:
@andy-armstrong @peter-fogg @andywaldrop @catong 

* agree that we don't need to do anything for the case where learners are in the discussion participating and you delete a team out from underneath them. 
* agree that we need some sort of confirmation after deletion to show that it worked, I think @marcotuts or @frrrances might have thoughts on that. 
* I'll let @catong review UI strings for deletion. I think we can say that discussions won't be accessible -- we don't need to say that we're not deleting them.

@peter-fogg I waited and the styling still seems screwed up on a few of the elements. The close button and the delete team / edit team membership seem too small and unlike other fonts in the LMS.

peter-fogg commented:
@explorerleslie I'm looking into the styling issues -- the prompt styling seems to be wrong throughout the LMS (see the "Leave Team" confirmation)>

catong commented:
@peter-fogg @explorerleslie @andy-armstrong How about this text for the delete modal:
"DELETE THIS TEAM? Deleting a team is permanent and cannot be undone. All members are removed from the team, and team discussions can no longer be accessed."

peter-fogg commented:
@andy-armstrong @frrrances @marcotuts Could someone take a look at the bad styling on the "delete team" warning? It looks like something went wrong when the system feedback framework was moved from `cms` to `common` -- this issue is present anywhere that one of these prompts is created in the LMS. I don't really have the SASS knowledge to be able to tell what the actual problem is, though.

cahrens commented on common/test/acceptance/pages/lms/teams.py:
```diff
@@ -229,10 +240,15 @@ def sort_teams_by(self, sort_order):
         ).click()
         self.wait_for_ajax()
 
+    def click_all_topics_link(self):
```
A "click all topics" method also exists in TeamsPage (which I would expect all the tests to have access to). I did not use the wait_for_ajax though, as I did not find it to be necessary.

cahrens commented on common/test/acceptance/pages/lms/teams.py:
```diff
@@ -138,6 +138,12 @@ def topic_cards(self):
         return self.q(css=TOPIC_CARD_CSS).results
 
     @property
+    def topic_team_counts(self):
```
Note that there is a verify_topic_team_count method in TeamsPage. 

I had to put some of these methods that are related to subview in TeamsPage because I wanted to test the refresh behavior after something has been marked stale. I found that using the more scoped objects caused them to be visited through their URLs, which meant that I couldn't reproduce the original bugs. I needed to navigate through the breadcrumbs and then verify things on the page without going through the routing URLs.

If you've found another way to do this that allows the methods to be in a more logical place, I'd be happy to have them moved. But I want to make sure that the tests fail without the "stale" logic in place.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -937,6 +949,97 @@ def test_user_can_cancel_the_team_creation(self):
+                staff=False,
+                roles=role
+            ).visit()
+        self.team_page.visit()
```
visit does wait_for_page, no need to call it explicitly.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -22,7 +22,15 @@
+    TeamManagementPage,
+    TeamPage
+)
+from ...pages.studio.utils import confirm_prompt
```
It would be good to move this to a more general util file, now that we have the same prompt being used by both LMS and CMS.

cahrens commented on lms/djangoapps/teams/static/teams/js/views/topic_teams.js:
```diff
@@ -16,7 +16,7 @@
                 initialize: function(options) {
                     TeamsView.prototype.initialize.call(this, options);
-                    _.bindAll(this, 'browseTeams', 'searchTeams', 'showCreateTeamForm');
+                    // _.bindAll(this, 'browseTeams', 'searchTeams', 'showCreateTeamForm');
```
Don't forget to delete this.

peter-fogg replied:
I ended up just removing these all in #9559.

cahrens commented on lms/djangoapps/teams/views.py:
```diff
@@ -541,6 +561,15 @@ def get_queryset(self):
+        """DELETE /api/team/v0/teams/{team_id}"""
+        team = get_object_or_404(CourseTeam, team_id=team_id)
+        self.check_object_permissions(request, team)
+        # Note: also deletes all team memberships associated with this team
```
We get this for free?

peter-fogg replied:
Yes, Django automatically propagates deletes to anything with a foreign key to the deleted model. https://docs.djangoproject.com/en/1.4/ref/models/querysets/#delete

cahrens commented on lms/djangoapps/teams/views.py:
```diff
@@ -530,6 +536,20 @@ class TeamsDetailView(ExpandableFieldViewMixin, RetrievePatchAPIView):
             method returns a 400 error with all error messages in the
             "field_errors" field of the returned JSON.
+
+        **Response Values for DELETE**
```
Did you also update the wiki spec?

peter-fogg replied:
Yes.

cahrens commented on lms/djangoapps/teams/views.py:
```diff
@@ -541,6 +561,15 @@ def get_queryset(self):
+    def delete(self, request, team_id):
+        """DELETE /api/team/v0/teams/{team_id}"""
+        team = get_object_or_404(CourseTeam, team_id=team_id)
+        self.check_object_permissions(request, team)
```
Do we actually need to call check_object_permissions explicitly? And why is the team passed to it?

peter-fogg replied:
Yeah, if I remove that line the tests around permissions to delete fail. https://tomchristie.github.io/rest-framework-2-docs/api-guide/permissions

cahrens replied:
OK, I see the difference now in how we do permissions most places. I agree this is nicer.

cahrens commented on lms/djangoapps/teams/static/teams/js/spec/views/instructor_tools_spec.js:
```diff
@@ -0,0 +1,71 @@
+    'teams/js/views/instructor_tools',
+    'teams/js/spec_helpers/team_spec_helpers',
+    'common/js/spec_helpers/ajax_helpers'
+], function ($, Backbone, Team, InstructorToolsView, TeamSpecHelpers, AjaxHelpers) {
```
Declare dependency on underscore.

cahrens commented on lms/djangoapps/teams/static/teams/js/spec/views/instructor_tools_spec.js:
```diff
@@ -0,0 +1,71 @@
+            };
+
+        beforeEach(function () {
+            setFixtures('<div id="page-prompt"></div>')
```
Missing semi-colon.

peter-fogg commented:
@cahrens I've addressed your feedback.

FYI @explorerleslie @andy-armstrong I've also added a confirmation that the team is deleted, as well as an error message if the deletion returns a 404. I didn't do anything about 401 or 403 (permission errors), since those will be handled in [TNL-3042](https://openedx.atlassian.net/browse/TNL-3042). The sandbox will be updated in a few minutes with the new code.

andy-armstrong commented on common/test/acceptance/pages/studio/container.py:
```diff
@@ -6,7 +6,8 @@
 
-from .utils import click_css, confirm_prompt, type_in_codemirror
+from .utils import type_in_codemirror
+from ..common.utils import click_css, confirm_prompt
```
Nit: I put ``..`` imports before ``.``, usually separated by a newline.

peter-fogg replied:
I'll clean these up.

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/collections/topic.js:
```diff
@@ -25,9 +25,7 @@
-                    if (event.action === 'create') {
-                        this.isStale = true;
-                    }
+                    this.isStale = this.isStale || _.contains(['create', 'delete'], event.action);
```
Nit: I'm finding this hard to read. How about:

    if (_.contains(['create', 'delete'], event.action)) {
        this.isStale = true;
    }

I think that better captures the logic, which is that we only mark the collection as stale for events of that action type.

peter-fogg replied:
I agree, that's much better. I'll fix that up.

andy-armstrong commented:
@peter-fogg Instead of holding this PR up, maybe we should make a new JIRA bug for the modal styling issue (since you aren't introducing it).

peter-fogg commented:
@andy-armstrong I'm all for it.

andy-armstrong commented:
@peter-fogg I found a minor refreshing bug. If you delete a team, then the message doesn't go away as you page through the teams, change the sort order etc. I think this is because it doesn't switch pages, but instead just updates the list view in place. You might want to explicitly remove any message in that situation too. I can show you what I mean if this isn't clear.

andy-armstrong commented:
:+1: great stuff!

andy-armstrong commented:
@peter-fogg I've created a JIRA ticket for the modal issue:

https://openedx.atlassian.net/browse/TNL-3189

peter-fogg commented:
@andy-armstrong I've addressed those nits and fixed the team message bug.

andy-armstrong commented:
:+1:

cahrens commented:
I find this styling (blue with a little green on top) odd:

![image](https://cloud.githubusercontent.com/assets/484484/9642785/2e717648-518c-11e5-8def-20a959083b82.png)

cahrens commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -554,14 +559,7 @@
+                    TeamUtils.showMessage(message);
                 },
 
                 hideWarning: function () {
```
hideWarning should be deleted.

cahrens commented on lms/djangoapps/teams/static/teams/js/views/topic_teams.js:
```diff
@@ -17,6 +17,7 @@
                 initialize: function(options) {
                     this.showSortControls = options.showSortControls;
                     TeamsView.prototype.initialize.call(this, options);
+                    _.bindAll(this, 'browseTeams', 'searchTeams', 'showCreateTeamForm');
```
Remove bindAll?

peter-fogg replied:
I've removed these all in #9559.

cahrens replied:
But this is adding it back in, isn't it?

cahrens commented:
A few minor things, otherwise :+1:

peter-fogg commented:
@cahrens The blue border on the success message is just the browser's outline for focused elements -- if you click elsewhere on the page it will go away. Putting focus on the message is necessary for accessibility.

catong commented:
@peter-fogg This looks great :+1:

catong commented:
@peter-fogg Oops and sorry -- I played around a bit more after giving thumbs, and realized that deleted teams show up in search. I created a story: TNL-3193.

explorerleslie commented:
@peter-fogg the modal styling still looks broken: 

![image](https://cloud.githubusercontent.com/assets/5643426/9658903/e43a04b6-521a-11e5-8b0c-353aa1127230.png)

And we should fix the deleted teams showing up in search results before merging this PR. Waiting to give my thumbs until then.

peter-fogg commented:
@explorerleslie We decided to break the modal styling out into a [separate story.](https://openedx.atlassian.net/browse/TNL-3189) I'm working on the search team bug right now; it shouldn't be too difficult to figure out.

peter-fogg commented:
@cahrens @bderusha The latest commit fixes deleted teams showing up in search; can you review?

cahrens commented:
:+1: Nice decorator!

explorerleslie commented:
:+1:

bderusha commented:
:+1:
